### ISSUE TITLE
invalid memory address or nil pointer dereference when mounting azure disk

### ISSUE THREAD

**Issue by ericewang** [kubernetes/kubernetes#55199]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
I deployed a spec using an AzureDisk that triggered an "invalid memory address or nil pointer dereference". The deployment was initializing a pod and mounting an AzureDisk volume after downloading the container. Here are the kubelet logs on the node during the time the error was thrown:
```
Nov 07 00:23:42 blue-minion-1 kubelet-wrapper[13163]: I1107 00:23:42.939539   13163 kubelet_node_status.go:499] Using Node Hostname from cloudprovider: "blue-minion-1"
Nov 07 00:23:45 blue-minion-1 kubelet-wrapper[13163]: I1107 00:23:45.265557   13163 qos_container_manager_linux.go:320] [ContainerManager]: Updated QoS cgroup configuration
Nov 07 00:23:50 blue-minion-1 kubelet-wrapper[13163]: I1107 00:23:50.011933   13163 kube_docker_client.go:330] Pulling image "prom/prometheus:v1.8.0": "12d20d1c8476: Pull complete "
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: I1107 00:23:52.308660   13163 kube_docker_client.go:333] Stop pulling image "prom/prometheus:v1.8.0": "Status: Downloaded newer 
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: E1107 00:23:52.310160   13163 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime erro
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /usr/local/go/src/runtime/asm_amd64.s:514
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /usr/local/go/src/runtime/panic.go:489
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /usr/local/go/src/runtime/panic.go:63
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /usr/local/go/src/runtime/signal_unix.go:290
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/azure_dd/azure_mounter.go:48
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_pods.go:138
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_pods.go:409
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:176
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:109
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:706
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1637
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:813
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:173
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:176
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:220
Nov 07 00:23:52 blue-minion-1 kubelet-wrapper[13163]: /usr/local/go/src/runtime/asm_amd64.s:2197
```

The line that throw the exception was [pkg/volume/azure_dd/azure_mounter.go:48](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/azure_mounter.go#L48), which accesses `volumeSource` when the `getVolumeSource` likely returned nil (in this [case](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/azure_common.go#L107)).

**What you expected to happen**:
An error message related to failed disk mounting is returned instead of a nil pointer deference.

**How to reproduce it (as minimally and precisely as possible)**:
Not very easy to reproduce, but we were in the process of moving pods between nodes running 1.5 -> 1.8. We suspect there may have been unexpected interactions that caused both `spec.Volume` and `spec.PersistentVolume` to be missing when the AzureDisk is fetched. Our pod spec uses AzureDisk volumes, not PersistentVolumes.

**Anything else we need to know?**:
I'd expect [this line](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/azure_mounter.go#L48) to not access `volumeSource` if an error was propagated up.

We worked around this issue by deleting the deployment and re-applying our spec, which did the trick, probably since the application of the config took place in an entirely 1.8 setting rather than a hybrid cluster (1.5 and 1.8 worker nodes).

**Environment**:
- Kubernetes version (use `kubectl version`): kubelet 1.8.1 & master running 1.8.1
- Cloud provider or hardware configuration: Azure
- OS (e.g. from /etc/os-release): CoreOS 1548.4.0
- Kernel (e.g. `uname -a`): Linux master 4.13.9-coreos #1 SMP Thu Oct 26 03:35:38 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz GenuineIntel GNU/Linux
- Install tools:
- Others:

**Comment by ericewang** [comment]
@kubernetes/sig-azure

**Comment by jianzj** [comment]
Hello @hzxuzhonghu 

Issue reported `nil` should be from `volumeSource, err := getVolumeSource(m.spec)` while should be `volumeSource` is nil, `err` equals `fmt.Errorf("azureDisk - Spec does not reference an Azure disk volume type")`

Maybe should we print this error into log ?

Thanks very much !

**Comment by hzxuzhonghu** [comment]
Make sense.

**Comment by andyzhangx** [comment]
@hzxuzhonghu thanks for the reporting, this issue is duplicate of https://github.com/kubernetes/kubernetes/issues/54149 which is fixed by https://github.com/kubernetes/kubernetes/pull/54607 and the fix is also in v1.8 now, while it's not in v1.8.1, line 48 in v1.8.1 is 
`ReadOnly:        *volumeSource.ReadOnly,`
